test/db_timeout_tests.c

```diff
@@ -0,0 +1,184 @@
+#include "greatest.h"
+
+#include <unistd.h>
+#include <sys/socket.h>
+
+#include "event_loop.h"
+#include "example_task.h"
+#include "common.h"
+#include "state/object_table.h"
+#include "state/redis.h"
+
+SUITE(db_timeout_tests);
+
+event_loop *loop;
+int simple_failed = 0;
+
+void simple_done_cb(object_id object_id,
+                    int manager_count,
+                    OWNER const char *manager_vector[],
+                    void *context) {
+  /* The done callback should not be called. */
+  CHECK(0);
+}
+
+void simple_fail_cb(unique_id id, void *user_data) {
+  simple_failed = 1;
+  CHECK(user_data == (void*) 42);
+  event_loop_stop(loop);
+}
+
+TEST timeout_test(void) {
+  loop = event_loop_create();
+  db_handle *db = db_connect("127.0.0.1", 6379, "plasma_manager", "127.0.0.1",
+                              1234);
+  db_attach(db, loop);
+  void *context = (void *) 42;
+  object_table_lookup(db, NIL_ID, 5, 100, simple_done_cb, simple_fail_cb, context);
+  /* Disconnect the database to see if the lookup times out. */
+  close(db->context->c.fd);
+  event_loop_run(loop);
+  db_disconnect(db);
+  event_loop_destroy(loop);
+  CHECK(simple_failed);
+  PASS();
+}
+
+int retry_succeeded = 0;
+
+void retry_done_cb(object_id object_id,
+                   int manager_count,
+                   OWNER const char *manager_vector[],
+                   void *context) {
+  CHECK(context == (void *) 42);
+  retry_succeeded = 1;
+  free(manager_vector);
+}
+
+void retry_fail_cb(unique_id id, void *user_data) {
+  CHECK(user_data == (void *) 42);
+  /* The fail callback should not be called. */
+  CHECK(0);
+}
+
+int64_t reconnect_db_cb(event_loop *loop, int64_t timer_id, void *context) {
+  db_handle *db = context;
+  /* Reconnect to redis. This is not reconnecting the pub/sub channel. */
+  redisAsyncFree(db->context);
+  redisFree(db->sync_context);
+  db->context = redisAsyncConnect("127.0.0.1", 6379);
+  db->context->data = (void *) db;
+  db->sync_context = redisConnect("127.0.0.1", 6379);
+  /* Re-attach the database to the event loop (the file descriptor changed). */
+  db_attach(db, loop);
+  return EVENT_LOOP_TIMER_DONE;
+}
+
+int64_t terminate_event_loop_cb(event_loop *loop, int64_t timer_id, void *context) {
+  event_loop_stop(loop);
+  return EVENT_LOOP_TIMER_DONE;
+}
+
+TEST retry_test(void) {
+  loop = event_loop_create();
+  db_handle *db = db_connect("127.0.0.1", 6379, "plasma_manager", "127.0.0.1", 11235);
+  db_attach(db, loop);
+  void *context = (void *) 42;
+  object_table_lookup(db, NIL_ID, 5, 100, retry_done_cb, retry_fail_cb, context);
```

Comment:
Unfortunately, I think this is really only a test for the object table timeout, not generic timeouts. This goes back to the other comment in state/table.c about making sure we clean up everything properly if `done_cb` gets called. Currently, the tables seem to do different things once they've called `done_cb`.